cmd/zdb/zdb.c

```diff
@@ -2363,6 +2338,96 @@ dump_nvlist_stats(nvlist_t *nvl, size_t cap)
 	nvlist_free(stats.zns_boolean);
 }
 
+static boolean_t
+buffer_is_unique(char *buf, size_t buflen, int l,
+    zio_cksum_t cksum[VDEV_LABELS], int *dupe_of)
+{
+	int prev_label;
+	boolean_t rc = B_TRUE;
+
+	fletcher_4_native_varsize(buf, buflen, &cksum[l]);
+
+	for (prev_label = 0; prev_label < l; prev_label++) {
+		if (memcmp(&cksum[l], &cksum[prev_label],
+		    sizeof (cksum[l])) == 0) {
+			*dupe_of = prev_label;
+			rc = B_FALSE;
+			break;
+		}
+	}
+
+	return (rc);
+}
+
+static void
+dump_config_from_label(nvlist_t *config, char *buf, int l,
+    zio_cksum_t cksum[VDEV_LABELS])
+{
+	size_t config_size;
+	boolean_t is_unique;
+	int dupe;
+	vdev_phys_t *vdev_phys;
+	size_t buflen = sizeof (vdev_phys->vp_nvlist);
+
+	/* do not checksum unused space in buf */
+	if (nvlist_size(config, &config_size, NV_ENCODE_XDR) != 0)
+		config_size = buflen;
+
+	if (!dump_opt['q']) {
+		/* dump config only if it differs from other labels */
+		is_unique = buffer_is_unique(buf, config_size, l, cksum,
+		    &dupe);
+		if (is_unique == B_TRUE) {
+			dump_nvlist(config, 4);
+			if (dump_opt['l'] >= 2)
+				dump_nvlist_stats(config, buflen);
+		} else {
+			(void) printf("    Configuration same as label %d\n",
+			    dupe);
+		}
+	}
+}
+
+#define	ZDB_MAX_UB_HEADER_SIZE 32
+#define	MAX_UBERBLOCK_COUNT	(VDEV_UBERBLOCK_RING >> UBERBLOCK_SHIFT)
+
+static void
+dump_label_uberblocks(vdev_label_t *lbl, uint64_t ashift, int label_num,
+    zio_cksum_t cksum[MAX_UBERBLOCK_COUNT][VDEV_LABELS])
+{
+	vdev_t vd;
+	vdev_t *vdp = &vd;
+	char header[ZDB_MAX_UB_HEADER_SIZE];
+	int i;
+	boolean_t is_unique;
+	int dupe;
+
+	vd.vdev_ashift = ashift;
+	vdp->vdev_top = vdp;
+
+	for (i = 0; i < VDEV_UBERBLOCK_COUNT(vdp); i++) {
```

Comment:
I initially thought of the all-or-nothing dump as well, but opted for this method to make the different one(s) stand out more.
I agree it should be simple to describe ranges that are the same and would reduce the output dramatically.  I'll do that.